sumcheck/src/prover_helper/arbitrary_gate.rs

```diff
@@ -0,0 +1,187 @@
+    /// - `r_eval_point`: Final random point of evaluation
+    /// - `final_evals`: Each of the polys evaluated at `r_eval_point`
+    pub fn prove_arbitrary<Func, T>(
+        _claim: &F,
+        num_rounds: usize,
+        polys: &mut [MultiLinearPoly<F>],
+        comb_func: Func,
+        combined_degree: usize,
+        transcript: &mut T,
+    ) -> (Self, Vec<F>, Vec<F>)
+    where
+        Func: Fn(&[F]) -> F + std::marker::Sync,
+        T: Transcript<F>,
+    {
+        let mut r: Vec<F> = Vec::new();
+        let mut uni_polys: Vec<UnivariatePolynomial<F>> = Vec::new();
+
+        for _round in 0..num_rounds {
+            let mle_half = polys[0].coeffs.len() / 2;
+
+            let accum: Vec<Vec<F>> = (0..mle_half)
+                .map(|poly_term_i| {
+                    let mut accum = vec![F::zero(); combined_degree + 1];
+                    // Evaluate P({0, ..., |g(r)|})
+
+                    // TODO(#28): Optimize
+                    // Tricks can be used here for low order bits {0,1} but general premise is a running sum for each
+                    // of the m terms in the Dense multilinear polynomials. Formula is:
+                    // half = | D_{n-1} | / 2
+                    // D_n(index, r) = D_{n-1}[half + index] + r * (D_{n-1}[half + index] - D_{n-1}[index])
+
+                    // eval 0: bound_func is A(low)
+                    let params_zero: Vec<F> =
+                        polys.iter().map(|poly| poly.coeffs[poly_term_i]).collect();
+                    accum[0] += comb_func(&params_zero);
+
+                    // TODO(#28): Can be computed from prev_round_claim - eval_point_0
+                    let params_one: Vec<F> = polys
+                        .iter()
+                        .map(|poly| poly.coeffs[mle_half + poly_term_i])
+                        .collect();
+                    accum[1] += comb_func(&params_one);
+
+                    // D_n(index, r) = D_{n-1}[half + index] + r * (D_{n-1}[half + index] - D_{n-1}[index])
+                    // D_n(index, 0) = D_{n-1}[LOW]
+                    // D_n(index, 1) = D_{n-1}[HIGH]
+                    // D_n(index, 2) = D_{n-1}[HIGH] + (D_{n-1}[HIGH] - D_{n-1}[LOW])
+                    // D_n(index, 3) = D_{n-1}[HIGH] + (D_{n-1}[HIGH] - D_{n-1}[LOW]) + (D_{n-1}[HIGH] - D_{n-1}[LOW])
+                    // ...
+                    let mut existing_term = params_one;
+                    for eval_i in accum.iter_mut().take(combined_degree + 1).skip(2) {
+                        let mut poly_evals = vec![F::zero(); polys.len()];
+                        for poly_i in 0..polys.len() {
+                            let poly = &polys[poly_i];
+                            poly_evals[poly_i] = existing_term[poly_i]
+                                + poly.coeffs[mle_half + poly_term_i]
+                                - poly.coeffs[poly_term_i];
+                        }
+
+                        *eval_i += comb_func(&poly_evals);
```

Comment:
please consider for syntax simplicity, that just use assign here? this is the same case with prior `accum[0]` and `accum[1]` that only visited once and assigned once, so please just assign rather than add to the value...